pkg/tools/prefix_pool/prefixpool.go

```diff
@@ -0,0 +1,642 @@
+
+	for i, j := 0, len(values)-1; i <= j; i, j = i+1, j-1 {
+		newValues[i] = values[j]
+		if i != j {
+			newValues[j] = values[i]
+		}
+	}
+	return newValues
+}
+
+func ReleasePrefixes(prefixes []string, released ...string) (remaining []string, err error) {
+	result := []string{}
+	if len(released) == 0 {
+		return prefixes, nil
+	}
+	result = removeDuplicates(append(prefixes, released...))
+
+	prefixByPrefixLen := map[int][]*net.IPNet{}
+
+	for _, prefix := range result {
+		_, ipnet, err := net.ParseCIDR(prefix)
+		if err != nil {
+			return nil, err
+		}
+		parentLen, _ := ipnet.Mask.Size()
+		nets := prefixByPrefixLen[parentLen]
+		nets = append(nets, ipnet)
+		prefixByPrefixLen[parentLen] = nets
+	}
+
+	for {
+		newPrefixByPrefixLen := map[int][]*net.IPNet{}
+		changes := 0
+		for basePrefix, value := range prefixByPrefixLen {
+			base := map[string]*net.IPNet{}
+
+			cvalue := append(newPrefixByPrefixLen[basePrefix], value...)
+			parentPrefix := []*net.IPNet{}
+			if len(cvalue) < 2 {
+				newPrefixByPrefixLen[basePrefix] = cvalue
+				continue
+			}
+			for _, netValue := range cvalue {
+				baseIp := &net.IPNet{
+					IP:   clearNetIndexInIP(netValue.IP, basePrefix),
+					Mask: netValue.Mask,
+				}
+				parentLen, addrLen := baseIp.Mask.Size()
+				bv := base[baseIp.String()]
+				if bv == nil {
+					base[baseIp.String()] = netValue
+				} else {
+					// We found item with same base IP, we we could join this two networks into one.
+					// Remove from current level
+					delete(base, baseIp.String())
+					// And put to next level
+					parentPrefix = append(parentPrefix, &net.IPNet{
+						IP:   baseIp.IP,
+						Mask: net.CIDRMask(parentLen-1, addrLen),
+					})
+					changes++
+				}
+			}
+			leftPrefixes := []*net.IPNet{}
+			// Put all not merged values
+			for _, value := range base {
+				leftPrefixes = append(leftPrefixes, value)
+			}
+			newPrefixByPrefixLen[basePrefix] = leftPrefixes
+			newPrefixByPrefixLen[basePrefix-1] = append(newPrefixByPrefixLen[basePrefix-1], parentPrefix...)
+		}
+		if changes == 0 {
+			// All is merged, we could exit
+			result = []string{}
+			for _, value := range newPrefixByPrefixLen {
+				for _, v := range value {
+					result = append(result, v.String())
+				}
+			}
+			return result, nil
+		}
+		prefixByPrefixLen = newPrefixByPrefixLen
+	}
+}
+
+func subnet(ipnet *net.IPNet, subnet_index int) (*net.IPNet, error) {
+	mask := ipnet.Mask
+
+	parentLen, addrLen := mask.Size()
+	newPrefixLen := parentLen + 1
+	if newPrefixLen > addrLen {
+		return nil, errors.Errorf("insufficient address space to extend prefix of %d", parentLen)
+	}
+
+	if uint64(subnet_index) > 2 {
+		return nil, errors.Errorf("prefix extension does not accommodate a subnet numbered %d", subnet_index)
+	}
+
+	return &net.IPNet{
+		IP:   setNetIndexInIP(ipnet.IP, subnet_index, newPrefixLen),
+		Mask: net.CIDRMask(newPrefixLen, addrLen),
+	}, nil
+}
+
+func AddressCount(prefixes ...string) uint64 {
```

Comment:
Can we make this thing private?

Reply:
We can even delete it, because it is used in tests only